ceph/crashes.go

```diff
@@ -0,0 +1,152 @@
+)
+
+// CrashesCollector collects information on how many crash reports are currently open.
+// These reports are counted by daemon/client name, and by status (new or archived).
+// This is NOT the same as new_crash_reports, that only counts new reports in the past
+// two weeks as reported by 'ceph health'.
+type CrashesCollector struct {
+	conn    Conn
+	logger  *logrus.Logger
+	version *Version
+
+	// We keep track of which daemons we've seen so that their error count
+	// can be reset to zero if the errors get purged.
+	knownEntities map[string]bool
+
+	CrashReports prometheus.GaugeVec
+}
+
+// NewCrashesCollector creates a new CrashesCollector instance
+func NewCrashesCollector(exporter *Exporter) *CrashesCollector {
+	labels := make(prometheus.Labels)
+	labels["cluster"] = exporter.Cluster
+
+	collector := &CrashesCollector{
+		conn:    exporter.Conn,
+		logger:  exporter.Logger,
+		version: exporter.Version,
+
+		knownEntities: map[string]bool{},
+
+		CrashReports: *prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace:   cephNamespace,
+				Name:        "crash_reports",
+				Help:        "Count of crashes reports per daemon, according to `ceph crash ls`",
+				ConstLabels: labels,
+			},
+			[]string{"daemon", "status"},
+		),
+	}
+
+	return collector
+}
+
+type crashEntry struct {
+	entity string
+	isNew  bool
+}
+
+// getCrashLs runs the 'crash ls' command and parses its results
+func (c *CrashesCollector) getCrashLs() ([]crashEntry, error) {
+	crashes := make([]crashEntry, 0)
+
+	// We parse the plain format because it is quite compact.
+	// The JSON output of this command is very verbose and might be too slow
+	// to process in an outage storm.
+	cmd, err := json.Marshal(map[string]interface{}{
+		"prefix": "crash ls",
+		"format": "plain",
+	})
+	if err != nil {
+		return crashes, err
+	}
+
+	buf, _, err := c.conn.MonCommand(cmd)
+	if err != nil {
+		return crashes, err
+	}
+
+	scanner := bufio.NewScanner(bytes.NewBuffer(buf))
+	for scanner.Scan() {
+		matched := crashLsLineRegex.FindStringSubmatch(scanner.Text())
+		if len(matched) == 3 {
+			crashes = append(crashes, crashEntry{matched[1], matched[2] == "*"})
+		} else if len(matched) == 2 {
+			// Just in case the line-end spaces were stripped
+			crashes = append(crashes, crashEntry{matched[1], false})
+		}
+	}
+
+	return crashes, nil
+}
+
+// processCrashLs takes the parsed results from getCrashLs and counts them
+// in a map. It also keeps track of which daemons we've see in the past, and
+// initializes all counts to zero where needed.
+func (c *CrashesCollector) processCrashLs(crashes []crashEntry) map[crashEntry]int {
+	crashMap := make(map[crashEntry]int)
+
+	for _, crash := range crashes {
+		c.knownEntities[crash.entity] = true
+	}
+	for entity := range c.knownEntities {
+		crashMap[crashEntry{entity, true}] = 0
+		crashMap[crashEntry{entity, false}] = 0
+	}
+	for _, crash := range crashes {
+		crashMap[crash]++
+	}
+
+	return crashMap
+}
```

Comment:
Why not use a ConstMetric here instead, avoiding the need to track historical state? (ceph_exporter should probably move to using ConstMetric entirely at some point.)

Reply:
Hmm I'm looking into what `ConstMetric` seems to be doing, and I'm not entirely sure that I understand. Looks like that would remove the series for entries we can't find instead of keeping known label combinations around?

Reply:
With `ConstMetric`, the prometheus library no longer stores any metrics for us. Each time `Collect()` is called, we simply emit the metrics we want. It's the best choice for exporters that are simply providing a "window" into data that can be queried each time.